build_scripts/M3_merged.yaml

```diff
@@ -0,0 +1,33 @@
+openapi: 3.0.0
+info:
+  title: M3_merged
+  version: 0.0.0
+  description: |
+    5GMS AS M3 API
+    Copyright © 2022 British Broadcasting Corporation
+    All rights reserved.
+tags:
+  - name: M3_ContentHostingProvisioning
+    description: '5G Media Streaming: Application Server Provisioning (M3) APIs: Content Hosting Provisioning'
+  - name: M3_ServerCertificatesProvisioning
+    description: '5G Media Streaming: Application Server Provisioning (M3) APIs: Server Certificates Provisioning'
+externalDocs:
+  description: 'TS 26.512 V16.x.x; 5G Media Streaming (5GMS); Protocols'
+  url: 'https://www.3gpp.org/ftp/Specs/archive/26_series/26.512/'
```

Comment:
Delete these lines?

Reply:
Is this manual merging of the two YAML files a requirement of the Python language binding plug-in for OpenAPI Generator?

What is the meaning of `~1` in the cross-references to the source YAML?

Reply:
This `M3_merged.yaml` is a manual merge/wrapper for the two M3 OpenAPI files so that openapi-generator can include both in its API module generation. The header is a merged version of the two M3 files (so that `externalDocs` part is in those files), it then includes all the paths from both files, but then references the individual files for the implementation details for each path.

The `~1` notation is an escape code for `/` in a node key name in the YAML hierarchy for the referenced part to include. A '/' in the reference is a YAML hierarchy path separator. This is required here because the path is a key name in the OpenAPI YAML and includes '/' characters which need to be escaped.

Reply:
See the [Swagger "Using $ref" page](https://swagger.io/docs/specification/using-ref/) for information `$ref` and the escape codes.

Reply:
Thanks for the link to the explanation of escaping in JSON Paths. Seems a bit weird. I have never seen forward slashes escaped in any other OpenAPI usage of `$ref`.

To take one of your examples, wouldn't `#/paths/~1certificates` expand to `#/paths//certificates`?

Reply:
On the merger of the two OpenAPI files, is this a limitation of the Python bindings produced by OpenAPI Generator? (I don't recall @devbbc having this constraint when generating C bindings for the Application Function.) Is there really no way for the Python code to import the two sets of bindings from the two M3 API definition files?

Assuming this merger does need to happen, shouldn't the merged YAML file be generated at build time from the two source YAML files rather than checking in a static merged file?

Reply:
`#/paths/~1certificates` would refer to:
```yaml
paths:
  /certificates:
     ...
```
It's the `/` in the `/certificates` key under paths that needs escaping.

Or to put it another way:
`$ref: #/paths/~1certificates` is expanded to the following path elements by splitting on `/`: `[docroot, unescape("paths"), unescape("~1certificates")]` => `[docroot, "paths", "/certificates"]`

The need for the merge file is a limit of openapi-generator on any API client/server interface for any language. Note however that the multiple API data models can co-exist from multiple runs of the generator, and it's the models that open5gs uses, not the client/server API. The client/server API is hand produced in open5gs to follow their state machine handling.

For the 5GMS Application Server, and to reduce development time, I've opted to use the generated client/server API functions in Python, hence the need to use a combined YAML file to generate the server API functions into one interface module.